csrc/id_model/indexing.cpp

```diff
@@ -120,6 +137,20 @@ std::tuple<std::vector<IterDomain*>, std::vector<Val*>> getAllocationDomains(
     }
   }
 
+  // Loop promotion may affect allocations. Promotions of intermediate
+  // domains may not be defined correctly. Only consider loop domains
+  // for now.
+  for (auto& allocation_domain : allocation_domains) {
+    bool is_loop = std::find(
+                       tv->getLeafDomain().begin(),
+                       tv->getLeafDomain().end(),
+                       allocation_domain) != tv->getLeafDomain().end();
+    if (!is_loop) {
+      continue;
+    }
+    allocation_domain = getLoopPromotion(allocation_domain, id_model);
```

Comment:
`isPartitionedLoop` uses the parallel type of `id` instead of the `getParallelType` of the loop group of `id`, can this be a problem? Similarly, in line 117 above, we are also using `loop_id->getParallelType()` instead of the parallel type of the loop group. IIRC, if the loop promotion id is replayed, we will not set its parallelization type.

Reply:
If I have
```
smem_tv[b, I1] ca_pos(1)
tv[TIDx, I1] = smem_tv[b, I1] + concrete_tv[I0, I1]
```
then should `smem_tv` be allocated as `[I0, I1]` or `[I1]`?

How about 
```
smem_tv[bTIDx, I1] ca_pos(1)
tv[TIDx, I1] = smem_tv[bTIDx, I1] + concrete_tv[I0, I1]
```
?

Reply:
Thank you. That's a real problem. We should either always look at a loop group or just propagate parallel types to all IDs, including promotion domains. I think the latter is a simpler solution. Will work on it.

Reply:
> If I have
> 
> ```
> smem_tv[b, I1] ca_pos(1)
> tv[TIDx, I1] = smem_tv[b, I1] + concrete_tv[I0, I1]
> ```
> 
> then should `smem_tv` be allocated as `[I0, I1]` or `[I1]`?
> 
> How about
> 
> ```
> smem_tv[bTIDx, I1] ca_pos(1)
> tv[TIDx, I1] = smem_tv[bTIDx, I1] + concrete_tv[I0, I1]
> ```
> 
> ?

In both cases, the allocation size should be `[I0, I1]`, right?

Reply:
Maybe not. In the first case, `[I1]` should be enough as long as a proper predicate is added.

Reply:
I think for both case, `[I1]` is enough. I think it's just a matter of whether `[I1]` or `[I0, I1]` is easier for analysis.

Reply:
I think what we currently do is we don't inline broadcast domains like `I0` of `smem_tv`. Let me check.

Reply:
This is what I was referring to:

https://github.com/NVIDIA/Fuser/blob/main/csrc/tensor_view.cpp#L167-L170

But that only applies to innermost broadcast domains, so in this case it is inlined.

Reply:
The current lowering only allocates `[I1]`. More specifically, when a domain is a broadcast domain, it's not allocated even when it's promoted. I'd keep this behavior as is.

Reply:
Looking at the code again, this should be already what happens (with #2371). Pure broadcast domains should be filtered out `!mayRequireAllocation`. I'll add a test.

Reply:
Added.